client/src/components/Ticketing/ticketingmanager/DiscountCodes/DiscountPopUp.tsx

```diff
@@ -0,0 +1,263 @@
+import React, {ReactElement, useEffect, useRef} from 'react';
+import {Field, Form} from 'react-final-form';
+import {FormInput} from '../../FormInput';
+import {DiscountCode} from './discountUtils';
+
+export interface DiscountPopUpProps {
+  errorMessage?: string;
+  onCancel: (event: any) => void;
+  onSubmit: (event: any) => void;
+  primaryLabel?: string;
+  title?: string;
+  values: DiscountCode;
+}
+
+/**
+ * The DiscountPopUp component handles adding or editing a discount code.
+ *
+ * @param {DiscountPopUpProps} props
+ * @param {string} props.errorMessage - Optional error message to display
+ * @param {func} props.onCancel - On cancel handler
+ * @param {func} props.onSubmit - On submit handler
+ * @param {string?} props.primaryLabel - Text inside primary button
+ * @param {string?} props.title - Title of discount code popup
+ * @param {Contact} props.values - The values to autofill the popup when it's rendered
+ * @returns {ReactElement} DiscountPopUp WonderTix form modal
+ */
+const DiscountPopUp = (props: DiscountPopUpProps): ReactElement => {
+  const {
+    errorMessage,
+    onCancel,
+    onSubmit,
+    primaryLabel = 'Submit',
+    title = 'Create Discount Code',
+    values,
+  } = props;
+
+  const discountCodePopUpRef = useRef(null);
+
+  useEffect(() => {
+    // Code within the useEffect traps focus within the DiscountPopUp when tabbing through
+    // interactible elements. See: https://medium.com/cstech/achieving-focus-trapping-in-a-react-modal-component-3f28f596f35b
+    const popUpElement = discountCodePopUpRef.current;
+    popUpElement.focus();
+
+    const focusableElements = popUpElement.querySelectorAll(
+      'button, input, [tabindex]:not([tabindex="-1"])',
+    );
+    const firstElement = focusableElements[0];
+    const lastElement = focusableElements[focusableElements.length - 1];
+
+    const handleTabKeyPress = (event) => {
+      if (event.key === 'Tab') {
+        if (event.shiftKey && document.activeElement === firstElement) {
+          event.preventDefault();
+          lastElement.focus();
+        } else if (!event.shiftKey && document.activeElement === lastElement) {
+          event.preventDefault();
+          firstElement.focus();
+        }
+      }
+    };
+
+    const handleEscapeKeyPress = (event) => {
+      if (event.key === 'Escape') {
+        if (onCancel) {
+          onCancel(event);
+        }
+      }
+    };
+
+    popUpElement.addEventListener('keydown', handleTabKeyPress);
+    popUpElement.addEventListener('keydown', handleEscapeKeyPress);
+
+    return () => {
+      popUpElement.removeEventListener('keydown', handleTabKeyPress);
+      popUpElement.removeEventListener('keydown', handleEscapeKeyPress);
+    };
+  }, []);
+
+  const validate = (formValues: DiscountCode) => {
+    const errors = {};
+    if (formValues.code?.length > 32) {
+      errors['code'] = 'Must be 32 characters or less';
+    }
+    if (formValues.amount < 0) {
```

Comment:
It might be worth checking to make sure at least one of amount/percent is supplied by the user.

Reply:
Good suggestion